test/basic.test.js

```diff
@@ -37,32 +37,25 @@ async function testQos (t, qos) {
   }
 
   var subscribers = []
-  var received = 0
 
   for (let i = 0; i < 10; i++) {
     subscribers.push(helper.startClient())
   }
 
   subscribers = await Promise.all(subscribers)
 
-  function onMessage (topic, message) {
-    if (topic === msg.topic) {
-      if (received++ < 10 || qos === 2) {
-        t.pass('message received')
-      }
-    }
-  }
-
-  for (let i = 0; i < subscribers.length; i++) {
-    const client = subscribers[i]
-    await client.subscribe(msg.topic)
-    client.once('message', onMessage)
-  }
+  // subscribe all subscribers
+  await Promise.all(subscribers.map(s => s.subscribe(msg.topic)))
```

Comment:
please use http://npm.im/p-map with some max concurrency. Otherwise this looks like a DoS and it will have a higher chance of timing out.